getzim.sh

```diff
@@ -0,0 +1,303 @@
+#!/bin/bash
+
+# internal
+
+BASE=$(readlink -f $(dirname "$0"))
+CACHE="$BASE/.cache"
+
+# conf
+
+BASEURL="https://download.kiwix.org/zim/"
+
+# ui funcs
+
+log() {
+  echo "$(date +%s): $*"
+}
+
+textmenu() { # allows the user to choose an id or item from a list
+  echo
+  a=($1) # turn $1 into an array
+  [ -z "${a[1]}" ] && echo "Skipping question \"$2\" because there is only one choice: ${a[0]}" && res="${a[0]}" && return 0
+  i=0
+  for c in $1; do # build a list
+    echo "[$i] $c"
+    i=$(expr $i + 1)
+  done
+  [ -z "$3" ] && read -p "[?] $2 > " _id #if no element was specified as cli arg ask the user
+  [ ! -z "$3" ] && _id="$3" #otherwise use that
+  id=$(echo "$_id" | sed "s|[^0-9]||g") #only numbers
+  for e in $1; do [ "$e" == "$_id" ] && res="$e" && echo "< $res" && return 0; done #check if item is in list
+  res=${a[$id]}
+  [ -z "$res" ] && [ ! -z "$3" ] && echo "Invalid ID or item: $3" && exit 2 #if id/item was specified via cli exit if invalid
+  [ -z "$id" ] && [ -z "$_id" ] && textmenu "$1" "$2" && return 0 #no input
+  [ -z "$id" ] && echo "Please enter a number or an item name" && textmenu "$1" "$2" && return 0
+  [ -z "$res" ] && echo "INVALID ID" && textmenu "$1" "$2" && return 0
+  echo "< $res" #show the choice to the user
+}
+
+# scraper fncs
+
+fetch_with_cache() {
+  OUTFILE=${1/"/"//"_"}
+  OUTFILE="$CACHE/_page$OUTFILE"
+
+  if [ -e "$OUTFILE" ]; then
+    cat "$OUTFILE"
+  else
+    OUT=$(curl -sL "$BASEURL$1")
+    echo "$OUT" > "$OUTFILE"
```

Comment:
When executing `url` without doing cache update first I got error:

```console
getzim.sh: line 49: /home/lidel/project/ipfs/distributed-wikipedia-mirror/.cache/_pagewikipedia: No such file or directory
```

Fix:

```suggestion
    mkdir -p "$CACHE"
    echo "$OUT" > "$OUTFILE"
```